sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/LDeviceAdapter.java

```diff
@@ -359,6 +378,59 @@ protected boolean hasControlBlockCreationCapability(ControlBlockEnum controlBloc
 
     private boolean hasCBNameConf(TServiceSettings tServiceSettings) {
         return tServiceSettings.isSetCbName()
-            && (TServiceSettingsNoDynEnum.CONF.equals(tServiceSettings.getCbName()));
+                && (TServiceSettingsNoDynEnum.CONF.equals(tServiceSettings.getCbName()));
     }
+
+    /**
+     * Update and/or create Monitoring LNs (LSVS and LGOS) into LDSUIED from ExtRefs binding
+     *
+     * @param tExtRefs              ExtRefs for which source Control Blocks (Goose or SMV) should be monitored
+     * @param monitoringLnClassEnum LNClass value for LN to monitor
+     * @return a list of SclReport Objects that contains errors
+     */
+    public Optional<SclReportItem> manageMonitoringLns(List<TExtRef> tExtRefs, String doName, MonitoringLnClassEnum monitoringLnClassEnum) {
+        if (tExtRefs.isEmpty()) {
+            return Optional.empty();
+        }
+        return getLNAdapters().stream().filter(lnAdapter -> monitoringLnClassEnum.value().equals(lnAdapter.getLNClass()))
+                .map(lnAdapter -> {
+                    Optional<SclReportItem> optionalSclReportItem = Optional.empty();
+                    ResumedDataTemplate filter = new ResumedDataTemplate(lnAdapter, doName, DA_SETSRCREF);
+                    Optional<ResumedDataTemplate> foundDai = lnAdapter.getDAI(filter, true).stream().findFirst();
+                    if (foundDai.isEmpty()) {
+                        optionalSclReportItem = Optional.of(SclReportItem.warning(lnAdapter.getXPath() + "/DOI@name=\"" + doName + "\"/DAI@name=\"setSrcRef\"/Val",
+                                "The DAI cannot be updated"));
+                    } else {
+                        ResumedDataTemplate daToUpdateFilter = foundDai.get();
+                        TLN lnToUpdate = lnAdapter.getCurrentElem();
+                        int i = 1;
+                        for (TExtRef tExtRef : tExtRefs) {
+                            if (i > 1) {
```

Comment:
I discovered after doing some test that we continue to add existing LGOS with the same `inst`, it is invalid
```shell
[org.xml.sax.SAXParseException; lineNumber: 0; columnNumber: 0; cvc-identity-constraint.4.1: Duplicate unique value [2,LGOS,] declared for identity constraint "uniqueLNInLDevice" of element "LDevice".]
javax.xml.bind.MarshalException: null
```
I suggest using `LNNAME = prefix+LNInst+LNClass` as the main key when adding new LN object

```suggestion
                           if(Collections.frequency(existingLNodeNames, lnAdapter.getLNodeName()) == 0) {
                                lnToUpdate = copySclElement(lnAdapter.getCurrentElem(), TLN.class);
                                getCurrentElem().getLN().add(lnToUpdate);
                                i++;
                            }
                            updateNewCreatedLnDaiValue(lnToUpdate, tExtRef, String.valueOf(i), daToUpdateFilter);
```


```java
// List of distinct LN objects based (LNNAME)
var existingLNodeNames =  getLNAdapters().stream().map(AbstractLNAdapter::getLNodeName).toList();
```

Reply:
Good remarks but this use case was not specified.
Issue is updated : we delete extrat LNs (LGOS , LSVS) and keep only first one before creating new ones from it